// regression/index.js
// Delta package: regression — interactive illustrations for the machine
// learning chapter.
//
// The file has two halves:
//
//   TOOLKIT (DeltaPlot) — a small 2D plotting library: scales, panels, grids,
//     point/curve/segment layers, drag & click interaction, controls. It knows
//     nothing about any particular illustration.
//
//   WIDGETS — the custom elements themselves, written against the toolkit.
//     Each one should read as *its own idea* and little else:
//       <delta-polynomial-interpolation>  Lagrange interpolation through
//                                         draggable nodes
//       <delta-overfitting>               degree slider, morphing least-squares
//                                         fit, accumulating train/test MSE
//       <delta-knn-example>               k-nearest-neighbour classification
//
// Everything is wrapped in one IIFE, so the only global this file creates is
// `DeltaPlot`. Delta inlines every package script into every page, and bare
// top-level names would collide with any other package sharing the document.
//
// Constraints worth knowing before editing:
//   - Classic browser script. No import/export — Delta inlines it as-is.
//   - No literal URL scheme anywhere, comments included: Delta's offline-safety
//     linter greps for one and warns that the output may not work offline.
//     Hence the SVG_NS join below.
//   - Attributes are read once, in connectedCallback. There is no
//     attributeChangedCallback; changing an attribute later does nothing.

(function () {
    "use strict";

    // =======================================================================
    //  TOOLKIT
    // =======================================================================

    // Built by joining so the source carries no literal URL scheme — that
    // substring would otherwise trip Delta's offline-safety linter, even though
    // a namespace URI is never fetched over the network.
    const SVG_NS = ["http", "www.w3.org/2000/svg"].join("://");

    // --- primitives ------------------------------------------------------

    function svg(tag, attrs, ...children) {
        const node = document.createElementNS(SVG_NS, tag);
        if (attrs) for (const k in attrs) if (attrs[k] != null) node.setAttribute(k, attrs[k]);
        for (const c of children) node.append(c);
        return node;
    }

    function el(tag, cls) {
        const node = document.createElement(tag);
        if (cls) node.className = cls;
        return node;
    }

    function fmt(v) {
        if (Math.abs(v) < 1e-9) return "0";
        return parseFloat(v.toFixed(2)).toString();
    }

    // A "nice" grid step (1/2/5 * 10^k) covering `raw`.
    function niceStep(raw) {
        const pow = Math.pow(10, Math.floor(Math.log10(raw)));
        const f = raw / pow;
        const nice = f < 1.5 ? 1 : f < 3 ? 2 : f < 7 ? 5 : 10;
        return nice * pow;
    }

    // Document-unique ids (clip paths, slider/label pairs). A counter rather
    // than Math.random so two instances on a page never collide by chance.
    let uidCount = 0;
    function uid(prefix) { return (prefix || "dp") + "-" + (++uidCount); }

    // Superscript digits, for log-axis labels like 10⁻².
    const SUP = { "-": "⁻", 0: "⁰", 1: "¹", 2: "²", 3: "³", 4: "⁴", 5: "⁵", 6: "⁶", 7: "⁷", 8: "⁸", 9: "⁹" };
    function sup(k) { return String(k).split("").map((ch) => SUP[ch] || ch).join(""); }

    // Localized UI string (falls back to the Portuguese default when Delta.t is
    // unavailable, e.g. in a bare test harness).
    function t(key, fallback) {
        return (typeof window !== "undefined" && window.Delta && window.Delta.t)
            ? window.Delta.t(key, fallback)
            : fallback;
    }

    // Parse a `points="x,y; x,y"` attribute. Returns null when nothing usable
    // is there, so callers can `|| generateDefault()`.
    function parsePoints(attr) {
        if (!attr) return null;
        const pts = attr.split(";").map((pair) => {
            const [x, y] = pair.split(",").map((s) => parseFloat(s.trim()));
            return { x, y };
        }).filter((p) => Number.isFinite(p.x) && Number.isFinite(p.y));
        return pts.length ? pts : null;
    }

    const MARGIN = { l: 46, r: 16, t: 16, b: 34 };

    // --- scales ----------------------------------------------------------
    //
    // A scale maps a data value to a pixel and back. Both directions are total:
    // `to` never returns NaN for a finite input, which matters because a
    // high-degree fit routinely evaluates to ±Infinity.
    //
    // There is no "invert the y axis" flag — you invert the *range*:
    //   y: { domain: [ymin, ymax], range: [bottomPx, topPx] }
    // which is what Panel does for you.

    function linear(spec) {
        const [d0, d1] = spec.domain;
        const [r0, r1] = spec.range;
        const span = d1 - d0;
        const clamp = !!spec.clamp;
        return {
            type: "linear", domain: [d0, d1], range: [r0, r1],
            to(v) {
                // A degenerate domain (a single degree, in the MSE panel) would
                // otherwise divide by zero and paint the whole panel NaN.
                let f = span === 0 ? 0 : (v - d0) / span;
                if (clamp) f = Math.max(0, Math.min(1, f));
                return r0 + f * (r1 - r0);
            },
            from(px) { return d0 + ((px - r0) / (r1 - r0)) * span; },
            ticks(n) {
                const lo = Math.min(d0, d1), hi = Math.max(d0, d1);
                if (!(hi > lo)) return [lo];
                const step = niceStep((hi - lo) / (n || 6));
                const start = Math.ceil(lo / step) * step;
                const out = [];
                // `start + i*step` rather than an accumulator: at 15+ ticks the
                // running sum drifts far enough to drop the last tick.
                for (let i = 0; start + i * step <= hi + 1e-9; i++) out.push(start + i * step);
                return out;
            },
        };
    }

    function log(spec) {
        const [d0, d1] = spec.domain;
        const [r0, r1] = spec.range;
        const clamp = spec.clamp !== false;
        const l0 = Math.log10(d0), l1 = Math.log10(d1);
        const span = l1 - l0;
        return {
            type: "log", domain: [d0, d1], range: [r0, r1],
            to(v) {
                // Two-sided clamp: floor the *value* at the low decade (log of a
                // non-positive number is not a number), then clamp the fraction.
                // Both are needed for the capped MSE axis to pin runaway error
                // to the top edge instead of flinging the dot off the panel.
                const val = clamp ? Math.max(v, d0) : v;
                let f = span === 0 ? 0 : (Math.log10(val) - l0) / span;
                if (!Number.isFinite(f)) f = 0;
                if (clamp) f = Math.max(0, Math.min(1, f));
                return r0 + f * (r1 - r0);
            },
            from(px) { return Math.pow(10, l0 + ((px - r0) / (r1 - r0)) * span); },
            ticks() {
                const out = [];
                for (let k = Math.ceil(l0 - 1e-9); k <= Math.floor(l1 + 1e-9); k++) out.push(Math.pow(10, k));
                return out;
            },
        };
    }

    function makeScale(spec, range) {
        const full = { domain: spec.domain, range: spec.range || range, clamp: spec.clamp };
        return spec.type === "log" ? log(full) : linear(full);
    }

    // --- distance metrics -------------------------------------------------
    //
    // Which one you want is a real choice, not a detail. The plot is not square
    // in data terms (a 640x400 box over, say, a 2.0 x 1.5 window), so screen
    // distance weights x more heavily than data distance does — enough to
    // change which points are "nearest" near a class boundary.

    const metric = {
        data: (a, b) => Math.hypot(a.x - b.x, a.y - b.y),
        screen: (panel) => (a, b) => Math.hypot(
            panel.x.to(a.x) - panel.x.to(b.x),
            panel.y.to(a.y) - panel.y.to(b.y),
        ),
    };

    // --- panel ------------------------------------------------------------
    //
    // Owns the <svg>, the margins, the two scales, the ordered layers and the
    // clip rect. One element may build several (the overfitting widget builds a
    // main plot and an MSE panel).
    //
    // Layer creation order IS paint order. Create the grid before the curves
    // before the points.

    class Panel {
        constructor(opt) {
            this.W = opt.width;
            this.H = opt.height;
            this.m = Object.assign({}, MARGIN, opt.margin);
            this.plotW = this.W - this.m.l - this.m.r;
            this.plotH = this.H - this.m.t - this.m.b;

            this.setDomain({ x: opt.x, y: opt.y });

            this.svg = svg("svg", {
                class: opt.class || "dp-svg",
                viewBox: `0 0 ${this.W} ${this.H}`,
                role: "img",
                "aria-label": opt.ariaLabel || null,
            });

            this._layers = new Map();
            this._order = [];
            this._justDragged = false;

            this.clipUrl = null;
            if (opt.clip) {
                const id = uid("dp-clip");
                const clip = svg("clipPath", { id });
                clip.append(svg("rect", { x: this.m.l, y: this.m.t, width: this.plotW, height: this.plotH }));
                this.svg.append(svg("defs", null, clip));
                this.clipUrl = `url(#${id})`;
            }
        }

        // Replace one or both scales in place — the MSE panel's log axis
        // re-bounds whenever the data is resampled.
        setDomain(spec) {
            if (spec.x) this.x = makeScale(spec.x, [this.m.l, this.m.l + this.plotW]);
            // Inverted range: data y grows upward, pixel y grows downward.
            if (spec.y) this.y = makeScale(spec.y, [this.m.t + this.plotH, this.m.t]);
            return this;
        }

        layer(name, tag, attrs) {
            let node = this._layers.get(name);
            if (!node) {
                node = svg(tag || "g", attrs || { class: "dp-" + name });
                this._layers.set(name, node);
                this._order.push(node);
                this.svg.append(node);
            }
            return node;
        }

        text(x, y, str, anchor, cls) {
            const node = svg("text", { class: cls || "dp-tick", x, y, "text-anchor": anchor });
            node.textContent = str;
            return node;
        }

        // Pointer event -> data coordinates, through the SVG's own CTM so it is
        // correct under any CSS scaling of the responsive viewBox.
        eventToData(evt) {
            const pt = this.svg.createSVGPoint();
            pt.x = evt.clientX;
            pt.y = evt.clientY;
            const p = pt.matrixTransform(this.svg.getScreenCTM().inverse());
            return { x: this.x.from(p.x), y: this.y.from(p.y) };
        }

        clampToDomain(p) {
            const [x0, x1] = this.x.domain, [y0, y1] = this.y.domain;
            return {
                x: Math.min(Math.max(x0, x1), Math.max(Math.min(x0, x1), p.x)),
                y: Math.min(Math.max(y0, y1), Math.max(Math.min(y0, y1), p.y)),
            };
        }

        // Keep a blowing-up curve finite (the clip hides the excursion).
        clampPy(v, k) {
            const pad = (k == null ? 1.5 : k) * this.plotH;
            const lo = this.m.t - pad, hi = this.m.t + this.plotH + pad;
            if (!Number.isFinite(v)) return hi;
            return Math.max(lo, Math.min(hi, v));
        }

        // A transparent rect over the plotting area that catches clicks. Sits
        // below every layer, so points and curves keep their own hit targets.
        background(opt) {
            opt = opt || {};
            this.bg = svg("rect", {
                class: opt.class || "dp-bg",
                x: this.m.l, y: this.m.t, width: this.plotW, height: this.plotH,
                fill: "transparent",
            });
            if (this._order.length) this.svg.insertBefore(this.bg, this._order[0]);
            else this.svg.append(this.bg);
            if (opt.onAdd) {
                this.bg.addEventListener("click", (e) => {
                    // The click that closes a drag must not also add a point.
                    if (this._justDragged) { this._justDragged = false; return; }
                    opt.onAdd(this.eventToData(e), e);
                });
            }
            return this.bg;
        }

        // One grid renderer for every axis in the package. The knobs exist
        // because the widgets genuinely differ, not to be general: `label`
        // returning null skips a tick's text (the MSE panel labels every 5th
        // degree), and `ticks` overrides the scale's own choice.
        drawGrid(opt) {
            opt = opt || {};
            const g = this.layer(opt.layer || "grid");
            g.replaceChildren();

            const C = Object.assign(
                { line: "dp-gridline", axis: "dp-axis", tick: "dp-tick", title: "dp-axis-title" },
                opt.classes,
            );
            const visible = opt.visible !== false;
            const xs = opt.x || {}, ys = opt.y || {};
            const top = this.m.t, bottom = this.m.t + this.plotH;
            const left = this.m.l, right = this.m.l + this.plotW;

            const xFormat = xs.format || fmt;
            const xLabel = xs.label || ((v) => xFormat(v));
            const xDy = xs.labelDy == null ? 16 : xs.labelDy;
            for (const v of (xs.ticks || this.x.ticks(xs.count || 6))) {
                const px = this.x.to(v);
                if (visible) g.append(svg("line", { class: xs.lineClass || C.line, x1: px, y1: top, x2: px, y2: bottom }));
                const str = xLabel(v);
                if (str != null) g.append(this.text(px, bottom + xDy, str, "middle", C.tick));
            }

            const yFormat = ys.format || fmt;
            const yLabel = ys.label || ((v) => yFormat(v));
            const yDx = ys.labelDx == null ? 8 : ys.labelDx;
            const yDy = ys.labelDy == null ? 4 : ys.labelDy;
            for (const v of (ys.ticks || this.y.ticks(ys.count || 5))) {
                const py = this.y.to(v);
                if (visible) g.append(svg("line", { class: ys.lineClass || C.line, x1: left, y1: py, x2: right, y2: py }));
                const str = yLabel(v);
                if (str != null) g.append(this.text(left - yDx, py + yDy, str, ys.anchor || "end", C.tick));
            }

            // Emphasised axes at x = 0 / y = 0 when in view.
            const zero = Object.assign({ vertical: true, horizontal: true }, opt.zero);
            if (zero.horizontal && contains(this.y.domain, 0)) {
                const py = this.y.to(0);
                g.append(svg("line", { class: C.axis, x1: left, y1: py, x2: right, y2: py }));
            }
            if (zero.vertical && contains(this.x.domain, 0)) {
                const px = this.x.to(0);
                g.append(svg("line", { class: C.axis, x1: px, y1: top, x2: px, y2: bottom }));
            }

            if (opt.axisTitle && opt.axisTitle.x) {
                g.append(this.text(left + this.plotW / 2, this.H - 6, opt.axisTitle.x, "middle", C.title));
            }
            return g;
        }

        points(opt) { return new PointLayer(this, opt); }
        curve(opt) { return new CurveLayer(this, opt); }
        segments(opt) { return new SegmentLayer(this, opt); }
    }

    function contains(domain, v) {
        return Math.min(domain[0], domain[1]) <= v && v <= Math.max(domain[0], domain[1]);
    }

    // --- point layer -------------------------------------------------------
    //
    // The data is an array of {x, y, ...} — plain objects, not DOM nodes. The
    // circles are derived from it and thrown away on every draw(); anything you
    // want to survive belongs on the datum.

    class PointLayer {
        constructor(panel, opt) {
            opt = opt || {};
            this.panel = panel;
            this.g = panel.layer(opt.layer || "points");
            this.data = opt.data || [];
            this.opt = opt;
            this._drag = null;
        }

        _attrs(p, i) {
            const o = this.opt;
            const a = {
                class: typeof o.class === "function" ? o.class(p, i) : (o.class || "dp-point"),
                cx: this.panel.x.to(p.x),
                cy: this.panel.y.to(p.y),
                r: typeof o.r === "function" ? o.r(p, i) : (o.r == null ? 6 : o.r),
            };
            // The grab cursor and the hover swell are affordances: they promise
            // the reader this point moves. Only a layer that actually wired a
            // drag gets to make that promise.
            if (this._drag) a.class += " dp-draggable";
            if (o.attrs) Object.assign(a, o.attrs(p, i));
            return a;
        }

        draw() {
            this.g.replaceChildren();
            this.data.forEach((p, i) => {
                const c = svg("circle", this._attrs(p, i));
                if (this.opt.title) {
                    const title = svg("title");
                    title.textContent = this.opt.title(p, i);
                    c.append(title);
                }
                if (this._drag) {
                    c.addEventListener("pointerdown", (e) => this._start(e, i));
                    if (this._drag.onRemove) {
                        c.addEventListener("dblclick", (e) => { e.stopPropagation(); this._drag.onRemove(i); });
                    }
                }
                this.g.append(c);
            });
            return this;
        }

        node(i) { return this.g.children[i]; }

        // Re-apply class and attributes to one existing circle. Used mid-drag,
        // where a full draw() would destroy the node holding pointer capture.
        refresh(i) {
            const c = this.node(i);
            if (!c) return;
            const a = this._attrs(this.data[i], i);
            // A null must REMOVE the attribute, not be skipped. svg() skipping a
            // null on a fresh node leaves it absent; skipping it here would leave
            // the previous value behind — so a point losing its label would keep
            // the colour of the label it just lost.
            for (const k in a) {
                if (a[k] == null) c.removeAttribute(k);
                else c.setAttribute(k, a[k]);
            }
        }

        place(i, p) {
            const c = this.node(i);
            if (!c) return;
            c.setAttribute("cx", this.panel.x.to(p.x));
            c.setAttribute("cy", this.panel.y.to(p.y));
        }

        draggable(opt) { this._drag = opt || {}; return this; }

        _start(e, i) {
            const o = this._drag;
            e.preventDefault();
            e.stopPropagation();
            const circle = e.currentTarget;
            circle.setPointerCapture(e.pointerId);

            const move = (ev) => {
                let p = this.panel.eventToData(ev);
                if (o.clamp !== false) p = this.panel.clampToDomain(p);
                // Merge rather than replace: a datum may carry a label or any
                // other widget state that must survive being dragged.
                this.data[i] = Object.assign({}, this.data[i], p);
                circle.setAttribute("cx", this.panel.x.to(p.x));
                circle.setAttribute("cy", this.panel.y.to(p.y));
                if (o.onMove) o.onMove(i, this.data[i]);
            };
            const up = (ev) => {
                circle.releasePointerCapture(ev.pointerId);
                circle.removeEventListener("pointermove", move);
                circle.removeEventListener("pointerup", up);
                this.panel._justDragged = true;
                if (o.onEnd) o.onEnd(i, this.data[i]);
            };
            circle.addEventListener("pointermove", move);
            circle.addEventListener("pointerup", up);
        }
    }

    // --- curve layer -------------------------------------------------------
    //
    // Samples a function on a fixed x-grid and emits a path. The grid is fixed
    // so that two curves (or the same curve at two degrees) always have the
    // same vertex count, which is what makes a per-vertex morph possible.

    class CurveLayer {
        constructor(panel, opt) {
            opt = opt || {};
            this.panel = panel;
            this.decimals = opt.decimals == null ? 2 : opt.decimals;
            this.path = panel.layer(opt.layer || "curve", "path", {
                class: opt.class || "dp-curve",
                fill: "none",
                "clip-path": opt.clip ? panel.clipUrl : null,
            });

            const n = opt.samples == null ? 400 : opt.samples;
            const [d0, d1] = panel.x.domain;
            this.xValues = new Array(n + 1);
            this.xPixels = new Array(n + 1);
            for (let i = 0; i <= n; i++) {
                const x = d0 + (i / n) * (d1 - d0);
                this.xValues[i] = x;
                this.xPixels[i] = panel.x.to(x);
            }
        }

        // Two non-finite policies, both genuinely wanted:
        //   "break" — end the subpath, so an asymptote leaves a gap (the
        //             interpolation widget, where the blow-up is the point)
        //   "clamp" — pin to just outside the plot, keeping a continuous vertex
        //             list that can still be tweened (the overfitting widget)
        _py(y, policy) {
            if (policy === "clamp") return this.panel.clampPy(this.panel.y.to(y));
            return Number.isFinite(y) ? this.panel.y.to(y) : null;
        }

        setFn(fn, opt) {
            const policy = (opt && opt.onNonFinite) || "break";
            const d = this.decimals;
            let out = "";
            let pen = false; // false => next vertex starts a new subpath
            for (let i = 0; i < this.xValues.length; i++) {
                const py = this._py(fn(this.xValues[i]), policy);
                if (py == null) { pen = false; continue; }
                out += (pen ? " L" : " M") + this.xPixels[i].toFixed(d) + " " + py.toFixed(d);
                pen = true;
            }
            this.path.setAttribute("d", out.trim());
            return this;
        }

        // Pixel y for every sample, for callers that tween the pixels directly.
        pixelsFor(fn, opt) {
            const policy = (opt && opt.onNonFinite) || "clamp";
            const arr = new Array(this.xValues.length);
            for (let i = 0; i < this.xValues.length; i++) {
                const py = this._py(fn(this.xValues[i]), policy);
                arr[i] = py == null ? this.panel.clampPy(NaN) : py;
            }
            return arr;
        }

        setPixels(ys) {
            const d = this.decimals;
            let out = "";
            for (let i = 0; i < ys.length; i++) {
                out += (i ? " L" : "M") + this.xPixels[i].toFixed(d) + " " + ys[i].toFixed(d);
            }
            this.path.setAttribute("d", out);
            return this;
        }

        clear() { this.path.setAttribute("d", ""); return this; }
    }

    // --- segment layer -----------------------------------------------------
    //
    // Straight lines in data space: neighbour links, residuals, error bars.
    // set() rebuilds from state — pass everything you want visible, every time.
    // Appending incrementally is what leaks.

    class SegmentLayer {
        constructor(panel, opt) {
            opt = opt || {};
            this.panel = panel;
            this.g = panel.layer(opt.layer || "links");
            this.class = opt.class || "dp-link";
        }

        set(list) {
            const p = this.panel;
            this.g.replaceChildren();
            for (const s of list) {
                this.g.append(svg("line", {
                    class: s.class || this.class,
                    x1: p.x.to(s.x1), y1: p.y.to(s.y1),
                    x2: p.x.to(s.x2), y2: p.y.to(s.y2),
                }));
            }
            return this;
        }

        // One point to many — a k-NN neighbourhood, a star plot.
        fan(from, tos) {
            return this.set(tos.map((to) => ({ x1: from.x, y1: from.y, x2: to.x, y2: to.y })));
        }

        clear() { this.g.replaceChildren(); return this; }
    }

    // --- animation ---------------------------------------------------------

    const reduceMotion = () => typeof window !== "undefined" && window.matchMedia
        && window.matchMedia("(prefers-reduced-motion: reduce)").matches;

    const easeOutCubic = (t) => 1 - Math.pow(1 - t, 3);

    // Tween an array of numbers, calling onFrame with the live values. Returns
    // a handle so an in-flight tween can be cancelled by the next one.
    function tween(from, to, opt, onFrame) {
        opt = opt || {};
        const cur = from.slice();
        if (opt.respectReducedMotion !== false && reduceMotion()) {
            for (let i = 0; i < to.length; i++) cur[i] = to[i];
            onFrame(cur, 1);
            return { cancel() {} };
        }
        const dur = opt.duration == null ? 300 : opt.duration;
        const ease = opt.ease || easeOutCubic;
        const start = performance.now();
        let raf = null;
        const step = (now) => {
            const p = Math.min(1, (now - start) / dur);
            const e = ease(p);
            for (let i = 0; i < to.length; i++) cur[i] = from[i] + (to[i] - from[i]) * e;
            onFrame(cur, p);
            raf = p < 1 ? requestAnimationFrame(step) : null;
        };
        raf = requestAnimationFrame(step);
        return { cancel() { if (raf) cancelAnimationFrame(raf); raf = null; } };
    }

    // --- controls ----------------------------------------------------------

    // `accent` names one of Delta's document accents (red, blue, …); it re-scopes
    // the accent tokens for that swatch so a legend entry matches a series that
    // carries the same data-accent, without either side naming a colour.
    function legend(items, cls) {
        const row = el("div", cls || "dp-legend");
        for (const it of items) {
            const chip = el("span", "dp-chip");
            const sw = el("span", "dp-swatch " + (it.class || "") + (it.dash ? " dp-swatch-dash" : ""));
            if (it.accent) sw.setAttribute("data-accent", it.accent);
            chip.append(sw, document.createTextNode(it.label));
            row.append(chip);
        }
        return row;
    }

    function controls() {
        const box = el("div");
        const bar = el("div", "dp-controls");
        box.append(bar);
        return {
            el: box,   // bar + any hint; mount this when you want both
            bar,       // just the bar; mount this to place it yourself
            readout() { const s = el("span", "dp-readout"); bar.append(s); return s; },
            button(label, fn) {
                const b = el("button", "dp-btn");
                b.type = "button";
                b.textContent = label;
                b.addEventListener("click", fn);
                bar.append(b);
                return b;
            },
            checkbox(label, checked, fn) {
                const wrap = el("label", "dp-toggle");
                const cb = el("input");
                cb.type = "checkbox";
                cb.checked = !!checked;
                cb.addEventListener("change", () => fn(cb.checked));
                wrap.append(cb, document.createTextNode(" " + label));
                bar.append(wrap);
                return cb;
            },
            slider(o) {
                const id = uid("dp-slider");
                const lab = el("label", "dp-range-label");
                const val = el("span", "dp-range-value");
                val.textContent = o.value;
                lab.append(document.createTextNode(o.label + " = "), val);
                lab.htmlFor = id;

                const input = el("input", "dp-slider");
                input.type = "range";
                input.id = id;                    // generated: two widgets on one
                input.min = o.min;                // page must not share a label
                input.max = o.max;
                input.step = o.step == null ? 1 : o.step;
                input.value = o.value;
                if (o.ariaLabel) input.setAttribute("aria-label", o.ariaLabel);
                input.addEventListener("input", () => o.onInput(parseFloat(input.value)));

                bar.append(lab, input);
                return { label: lab, input, value: val };
            },
            legend(items, cls) { const r = legend(items, cls); bar.append(r); return r; },
            hint(text) {
                const p = el("p", "dp-hint");
                p.textContent = text;
                box.append(p);
                return p;
            },
            center() { bar.style.justifyContent = "center"; },
        };
    }

    // --- widget base -------------------------------------------------------

    class Widget extends HTMLElement {
        connectedCallback() {
            // <delta-interactive> re-parents its children into a frame div,
            // which disconnects and reconnects us. Without this guard every
            // framed widget builds itself twice.
            if (this.dataset.deltaReady === "1") return;
            this.dataset.deltaReady = "1";
            this.setup();
        }

        num(name, dflt) { const v = parseFloat(this.getAttribute(name)); return Number.isFinite(v) ? v : dflt; }
        int(name, dflt) { const v = parseInt(this.getAttribute(name), 10); return Number.isFinite(v) ? v : dflt; }
        flag(name, dflt) { const a = this.getAttribute(name); return a == null ? dflt : a !== "false"; }

        // Read xmin/xmax/ymin/ymax with per-widget defaults.
        domain(def) {
            return {
                x: [this.num("xmin", def.x[0]), this.num("xmax", def.x[1])],
                y: [this.num("ymin", def.y[0]), this.num("ymax", def.y[1])],
            };
        }

        mount(...nodes) {
            this.replaceChildren();
            const wrap = el("div", "dp-wrap");
            wrap.append(...nodes);
            this.append(wrap);
            return wrap;
        }
    }

    function define(tag, cls) {
        if (!customElements.get(tag)) customElements.define(tag, cls);
    }

    const P = {
        svg, el, fmt, niceStep, uid, sup, t, parsePoints, MARGIN,
        linear, log, metric, Panel, tween, easeOutCubic, reduceMotion,
        controls, legend, Widget, define,
    };

    // =======================================================================
    //  WIDGETS
    // =======================================================================

    // Runge's function 1/(1 + 25x^2) — the classic example of interpolation
    // blowing up (oscillating) as the number of equally spaced nodes grows.
    const runge = (x) => 1 / (1 + 25 * x * x);

    // -----------------------------------------------------------------------
    //  <delta-polynomial-interpolation>
    //
    //  Points can be dragged, added (click) and removed (double-click); the
    //  unique interpolating polynomial of degree <= n-1 is drawn through them.
    // -----------------------------------------------------------------------

    // Barycentric weights for a set of (distinct) x-nodes. Reused across every
    // sample of the interpolant, so evaluation is O(n) per point after this O(n^2).
    function barycentricWeights(xs) {
        const n = xs.length;
        const w = new Array(n).fill(1);
        for (let j = 0; j < n; j++) {
            for (let k = 0; k < n; k++) {
                if (k === j) continue;
                const d = xs[j] - xs[k];
                w[j] /= d === 0 ? 1e-9 : d;
            }
        }
        return w;
    }

    // Barycentric form of the Lagrange interpolant (numerically stable).
    function evalInterp(xs, ys, w, x) {
        let num = 0, den = 0;
        for (let j = 0; j < xs.length; j++) {
            const diff = x - xs[j];
            if (Math.abs(diff) < 1e-12) return ys[j]; // x hits a node exactly
            const t = w[j] / diff;
            num += t * ys[j];
            den += t;
        }
        return num / den;
    }

    class PolynomialInterpolation extends P.Widget {
        setup() {
            const d = this.domain({ x: [-1, 1], y: [-0.3, 1.2] });
            this.showRunge = this.flag("runge", true);

            const panel = this.panel = new P.Panel({
                width: 640, height: 400, margin: P.MARGIN,
                x: { domain: d.x }, y: { domain: d.y },
                class: "dp-svg dp-interactive",
                ariaLabel: P.t("pi.aria", "Ilustração interativa de interpolação polinomial"),
                clip: true,
            });

            this.points = P.parsePoints(this.getAttribute("points")) || this._rungeNodes(8);

            panel.drawGrid({ visible: this.flag("grid", true) });
            this.ghost = panel.curve({ layer: "ghost", class: "dp-curve-ghost", clip: true, samples: 400 });
            this.curve = panel.curve({ layer: "curve", class: "dp-curve", clip: true, samples: 400 });
            this.pts = panel.points({
                data: this.points, r: 6,
                attrs: (p, i) => ({ "data-i": i, tabindex: 0 }),
                title: (p) => `(${P.fmt(p.x)}, ${P.fmt(p.y)}) — ` +
                    P.t("pi.point_hint", "arraste para mover, duplo-clique remove"),
            });
            this.pts.draggable({
                onMove: () => this._drawCurves(),   // cheap: the curve only
                onEnd: () => this.update(),
                onRemove: (i) => { this.points.splice(i, 1); this.update(); },
            });
            panel.background({ onAdd: (p) => { this.points.push(p); this.update(); } });

            this.mount(panel.svg, this._controls());
            this.update();
        }

        _rungeNodes(n) {
            const [ymin, ymax] = this.panel ? this.panel.y.domain : [this.num("ymin", -0.3), this.num("ymax", 1.2)];
            const [xmin, xmax] = this.panel ? this.panel.x.domain : [this.num("xmin", -1), this.num("xmax", 1)];
            const amp = (ymax - ymin) / 10;   // mean-zero uniform noise, <= a tenth of the range
            const pts = [];
            for (let i = 0; i < n; i++) {
                const x = xmin + (i / (n - 1)) * (xmax - xmin);
                pts.push({ x, y: runge(x) + (2 * Math.random() - 1) * amp });
            }
            return pts;
        }

        _setPoints(pts) {
            this.points = pts;
            this.pts.data = pts;
            this.update();
        }

        _controls() {
            const ui = P.controls();
            this.readout = P.el("span", "dp-readout");

            if (this.getAttribute("controls") === "false") {
                ui.button(P.t("pi.clear", "Limpar"), () => this._setPoints([]));
                ui.center();
            } else {
                ui.bar.append(this.readout);
                ui.button(P.t("pi.runge8", "Runge · 8 nós"), () => this._setPoints(this._rungeNodes(8)));
                ui.button(P.t("pi.runge14", "Runge · 14 nós"), () => this._setPoints(this._rungeNodes(14)));
                ui.button(P.t("pi.clear", "Limpar"), () => this._setPoints([]));
                ui.checkbox(P.t("pi.runge_toggle", "função de Runge"), this.showRunge, (on) => {
                    this.showRunge = on;
                    this.update();
                });
            }
            ui.hint(P.t("pi.hint", "Clique no plano para adicionar um nó · arraste para mover · duplo-clique remove."));
            return ui.el;
        }

        _drawCurves() {
            // Distinct x-nodes only — interpolation is undefined for repeated x.
            const seen = new Set();
            const nodes = [];
            for (const p of this.points) {
                const key = p.x.toFixed(6);
                if (seen.has(key)) continue;
                seen.add(key);
                nodes.push(p);
            }
            nodes.sort((a, b) => a.x - b.x);

            if (nodes.length >= 2) {
                const xs = nodes.map((p) => p.x);
                const ys = nodes.map((p) => p.y);
                const w = barycentricWeights(xs);
                this.curve.setFn((x) => evalInterp(xs, ys, w, x));
            } else {
                this.curve.clear();
            }

            if (this.showRunge) this.ghost.setFn(runge); else this.ghost.clear();
        }

        _updateReadout() {
            const n = this.points.length;
            this.readout.textContent = n === 0
                ? P.t("pi.empty", "Sem nós — clique no plano para começar.")
                : n === 1
                    ? P.t("pi.one", "1 nó — adicione mais para traçar o polinômio.")
                    : `n = ${n} nós → polinômio de grau ≤ ${n - 1}`;
        }

        update() {
            this._drawCurves();
            this.pts.draw();
            this._updateReadout();
        }
    }

    P.define("delta-polynomial-interpolation", PolynomialInterpolation);

    // -----------------------------------------------------------------------
    //  <delta-overfitting>
    //
    //  A dashed "unknown" function f (Runge) is sampled into noisy TRAINING
    //  points (accent) and an independent TEST set (a fixed contrasting hue). A
    //  slider sweeps the polynomial degree 0->maxdegree; the least-squares fit
    //  MORPHS smoothly between degrees (each vertex is tweened, so the curve
    //  visibly curls more rather than flashing), and a second panel accumulates
    //  the train/test MSE as the degree grows — the textbook "train MSE down,
    //  test MSE U-shape" picture.
    // -----------------------------------------------------------------------

    // Least-squares polynomial fit through Forsythe's discrete orthogonal
    // polynomials (three-term recurrence over the sample x's). Building the basis
    // degree by degree makes evaluation stable to high degree — where solving the
    // normal equations on a Vandermonde matrix would be hopelessly ill-conditioned.
    //   phi_0 = 1,  phi_{k+1}(x) = (x - a_k) phi_k(x) - b_k phi_{k-1}(x)  (b_0 = 0)
    //   a_k = <x phi_k, phi_k> / <phi_k, phi_k>,   b_k = <phi_k,phi_k>/<phi_{k-1},phi_{k-1}>
    //   c_k = <y, phi_k> / <phi_k, phi_k>          fit_d(x) = sum_{k<=d} c_k phi_k(x)
    function polyLeastSquares(xs, ys, maxDeg) {
        const n = xs.length;
        maxDeg = Math.max(0, Math.min(maxDeg, n - 1));
        const alpha = new Array(maxDeg + 1).fill(0);
        const beta = new Array(maxDeg + 1).fill(0);
        const coef = new Array(maxDeg + 1).fill(0);
        let pPrev = new Array(n).fill(0); // phi_{k-1} at each sample
        let pCur = new Array(n).fill(1);  // phi_k     at each sample (start: phi_0 = 1)
        let normPrev = 0;
        let usable = 0;
        for (let k = 0; k <= maxDeg; k++) {
            let normk = 0, xww = 0, yw = 0;
            for (let i = 0; i < n; i++) {
                const p = pCur[i];
                normk += p * p;
                xww += xs[i] * p * p;
                yw += ys[i] * p;
            }
            if (normk < 1e-12) { usable = k - 1; break; } // basis exhausted (repeated x's)
            usable = k;
            alpha[k] = xww / normk;
            coef[k] = yw / normk;
            beta[k] = k === 0 ? 0 : normk / normPrev;
            if (k < maxDeg) {
                const pNext = new Array(n);
                for (let i = 0; i < n; i++) pNext[i] = (xs[i] - alpha[k]) * pCur[i] - beta[k] * pPrev[i];
                pPrev = pCur; pCur = pNext; normPrev = normk;
            }
        }
        return {
            maxDeg: Math.max(0, usable),
            // Value of the degree-d least-squares fit at an arbitrary x.
            eval(x, d) {
                d = Math.max(0, Math.min(d, this.maxDeg));
                let phiPrev = 0, phiCur = 1, s = coef[0];
                for (let k = 1; k <= d; k++) {
                    const phi = (x - alpha[k - 1]) * phiCur - beta[k - 1] * phiPrev;
                    s += coef[k] * phi;
                    phiPrev = phiCur; phiCur = phi;
                }
                return s;
            },
        };
    }

    class Overfitting extends P.Widget {
        setup() {
            const d = this.domain({ x: [-1, 1], y: [-0.4, 1.3] });
            this.xmin = d.x[0]; this.xmax = d.x[1];

            const fixed = this.num("xfixed", null);
            this.xfixed = Number.isFinite(fixed) ? fixed : null;
            this.ntrain = Math.max(3, this.int("ntrain", 25));
            this.ntest = Math.max(2, this.int("ntest", 15));
            this.noiseAmp = this.num("noise", (d.y[1] - d.y[0]) / 10);
            // A degree-maxdeg fit must stay overdetermined (n_train > maxdeg) to
            // show genuine overfitting wiggle rather than exact interpolation.
            this.maxdeg = Math.max(1, Math.min(this.int("maxdegree", 15), this.ntrain - 2));
            this.degree = Math.max(0, Math.min(this.int("degree", 3), this.maxdeg));
            this.showGrid = this.flag("grid", true);
            // The TEST series re-scopes the accent tokens to one of Delta's own
            // document accents rather than naming a colour. Train keeps the
            // document accent, so the two only collide if the document itself is
            // set to this hue — hence the attribute, not a constant.
            this.testAccent = this.getAttribute("test-accent") || "blue";
            this._tween = null;

            this._generate();

            this.main = new P.Panel({
                width: 640, height: 340, margin: { l: 44, r: 14, t: 12, b: 28 },
                x: { domain: d.x }, y: { domain: d.y },
                class: "dp-svg of-main",
                ariaLabel: P.t("of.aria", "Ilustração interativa de sobreajuste em regressão polinomial"),
                clip: true,
            });
            this.mse = new P.Panel({
                width: 640, height: 208, margin: { l: 52, r: 14, t: 12, b: 36 },
                x: { domain: [0, this.maxdeg] },
                y: { type: "log", domain: this._mseDomain(), clamp: true },
                class: "dp-svg of-mse",
                ariaLabel: P.t("of.mse_aria", "MSE de treino e de teste em função do grau"),
            });

            this.main.drawGrid(this._mainGrid());
            this.truth = this.main.curve({ layer: "truth", class: "of-truth", clip: true, samples: 220, decimals: 1 });
            this.fitCurve = this.main.curve({ layer: "fit", class: "of-fit", clip: true, samples: 220, decimals: 1 });
            this.pts = this.main.points({
                data: [], r: 4,
                class: (p) => "dp-dot " + (p.kind === "test" ? "of-test" : p.x === this.xfixed ? "of-fixed" : "of-train"),
                attrs: (p) => (p.kind === "test" ? { "data-accent": this.testAccent } : null),
            });

            const ui = P.controls();
            this.slider = ui.slider({
                min: 0, max: this.maxdeg, value: this.degree,
                label: P.t("of.degree", "grau"),
                ariaLabel: P.t("of.degree_aria", "Grau do polinômio"),
                onInput: (v) => this._setDegree(v, true),
            });
            // Delta core's own .btn, not the package's .dp-btn — this widget has
            // always used the document button style; the other two have not.
            ui.button(P.t("of.resample", "novo conjunto"), () => this._resample()).className = "btn of-resample";

            this.mount(
                this.main.svg,
                P.legend([
                    { class: "of-truth", label: P.t("of.truth", "f (desconhecida)"), dash: true },
                    { class: "of-train", label: P.t("of.train", "treino") },
                    { class: "of-test", label: P.t("of.test", "teste"), accent: this.testAccent },
                ], "dp-legend of-legend"),
                ui.bar,
                this._mseHead(),
                this.mse.svg,
            );

            this.truth.setFn(runge, { onNonFinite: "clamp" });
            this._drawPoints();
            this.curveY = this.fitCurve.pixelsFor((x) => this.fit.eval(x, this.degree));
            this.fitCurve.setPixels(this.curveY);
            this._drawMseGrid();
            this._drawMse();
        }

        // --- data ---
        _sample(count) {
            const pts = new Array(count);
            for (let i = 0; i < count; i++) {
                const x = this.xmin + Math.random() * (this.xmax - this.xmin);
                const eps = (Math.random() * 2 - 1) * this.noiseAmp; // mean-zero uniform
                pts[i] = { x, y: runge(x) + eps };
            }
            if (this.xfixed != null && !pts.some((p) => p.x === this.xfixed)) {
                const eps = (Math.random() * 2 - 1) * this.noiseAmp;
                pts.push({ x: this.xfixed, y: runge(this.xfixed) + eps });
            }
            return pts;
        }

        _mseOf(set, d) {
            let s = 0;
            for (const p of set) { const e = p.y - this.fit.eval(p.x, d); s += e * e; }
            return s / set.length;
        }

        _generate() {
            this.train = this._sample(this.ntrain);
            this.test = this._sample(this.ntest);
            this.fit = polyLeastSquares(this.train.map((p) => p.x), this.train.map((p) => p.y), this.maxdeg);

            // Precompute train/test MSE for every degree (panel + axis bounds).
            this.mseTrain = new Array(this.maxdeg + 1);
            this.mseTest = new Array(this.maxdeg + 1);
            let lo = Infinity, hi = -Infinity;
            for (let d = 0; d <= this.maxdeg; d++) {
                this.mseTrain[d] = this._mseOf(this.train, d);
                this.mseTest[d] = this._mseOf(this.test, d);
                lo = Math.min(lo, this.mseTrain[d], this.mseTest[d]);
                hi = Math.max(hi, this.mseTrain[d], this.mseTest[d]);
            }
            lo = Math.max(lo, 1e-6);
            this.logMin = Math.floor(Math.log10(lo));
            this.logMax = Math.ceil(Math.log10(hi));
            if (this.logMax <= this.logMin) this.logMax = this.logMin + 1;
            // Cap the visible span: a degree-15 fit can send test MSE to the moon,
            // which would squash the informative low-degree region into a sliver.
            // Bound the axis to ~6 decades; anything beyond pins to the top edge
            // (the log scale clamps), reading as "the error shoots off the chart".
            this.logMax = Math.min(this.logMax, this.logMin + 6);
        }

        _mseDomain() { return [Math.pow(10, this.logMin), Math.pow(10, this.logMax)]; }

        // --- drawing ---
        _mainGrid() {
            return {
                visible: this.showGrid,
                x: { labelDy: 15 },
                zero: { vertical: false },   // only the y = 0 rule, as before
            };
        }

        _drawPoints() {
            // Test first so training points (the ones the fit chases) sit on top.
            this.pts.data = [
                ...this.test.map((p) => ({ x: p.x, y: p.y, kind: "test" })),
                ...this.train.map((p) => ({ x: p.x, y: p.y, kind: "train" })),
            ];
            this.pts.draw();
        }

        _drawMseGrid() {
            const degrees = [];
            for (let d = 0; d <= this.maxdeg; d++) degrees.push(d);
            this.mse.drawGrid({
                visible: this.showGrid,
                x: {
                    ticks: degrees,
                    lineClass: "dp-gridline dp-gridline-faint",
                    labelDy: 15,
                    // Label every fifth degree plus the last, so the axis stays legible.
                    label: (d) => (d % 5 === 0 || d === this.maxdeg) ? String(d) : null,
                },
                y: { format: (v) => "10" + P.sup(Math.round(Math.log10(v))) },
                zero: { vertical: false, horizontal: false },
                axisTitle: { x: P.t("of.degree", "grau") },
            });
        }

        // Reveal MSE points for degrees 0..current — the set grows as the slider rises.
        _drawMse() {
            const g = this.mse.layer("data", "g", { class: "of-mse-data" });
            g.replaceChildren();
            const d = this.degree;
            const series = (arr, mod, accent) => {
                if (d >= 1) {
                    let pts = "";
                    for (let k = 0; k <= d; k++) {
                        pts += this.mse.x.to(k).toFixed(1) + "," + this.mse.y.to(arr[k]).toFixed(1) + " ";
                    }
                    g.append(P.svg("polyline", {
                        class: "of-mse-line " + mod, points: pts.trim(), fill: "none",
                        "data-accent": accent,
                    }));
                }
                for (let k = 0; k <= d; k++) {
                    const cur = k === d;
                    g.append(P.svg("circle", {
                        class: "of-mse-dot " + mod + (cur ? " of-dot-new" : ""),
                        cx: this.mse.x.to(k), cy: this.mse.y.to(arr[k]), r: cur ? 4.5 : 3,
                        "data-accent": accent,
                    }));
                }
            };
            series(this.mseTrain, "of-train", null);   // null => the document accent
            series(this.mseTest, "of-test", this.testAccent);
        }

        _mseHead() {
            const h = P.el("div", "of-mse-head");
            const title = P.el("span", "of-mse-title");
            title.textContent = P.t("of.mse_title", "Erro quadrático médio (MSE) por grau");
            h.append(title, P.legend([
                { class: "of-train", label: P.t("of.mse_train", "treino") },
                { class: "of-test", label: P.t("of.mse_test", "teste"), accent: this.testAccent },
            ], "dp-legend of-mse-legend"));
            return h;
        }

        // --- interaction ---
        _morphTo(target) {
            if (this._tween) this._tween.cancel();
            this._tween = P.tween(this.curveY, target, {}, (cur) => {
                this.curveY = cur;
                this.fitCurve.setPixels(cur);
            });
        }

        _setDegree(d, animate) {
            d = Math.max(0, Math.min(d, this.maxdeg));
            this.degree = d;
            this.slider.value.textContent = d;
            const target = this.fitCurve.pixelsFor((x) => this.fit.eval(x, d));
            if (!animate || !this.curveY) {
                if (this._tween) this._tween.cancel();
                this.curveY = target;
                this.fitCurve.setPixels(target);
            } else {
                this._morphTo(target);
            }
            this._drawMse();
        }

        _resample() {
            this._generate();
            this._drawPoints();
            this.mse.setDomain({ y: { type: "log", domain: this._mseDomain(), clamp: true } });
            this._drawMseGrid();   // log-axis bounds may have shifted
            this._morphTo(this.fitCurve.pixelsFor((x) => this.fit.eval(x, this.degree)));
            this._drawMse();
        }
    }

    P.define("delta-overfitting", Overfitting);

    // -----------------------------------------------------------------------
    //  <delta-knn-example>
    //
    //  Two clouds of labelled training points. Click to drop a query point; it
    //  links to its k nearest neighbours and takes their majority label. Drag it
    //  across the boundary and watch the label flip.
    //
    //  A query point with label === null is PENDING: the reader has placed it but
    //  not asked for a verdict. It draws hollow and contributes no links. That is
    //  the state the two classify toggles switch into.
    //
    //  Attributes split into what the widget DOES and what the reader may CHANGE:
    //    behaviour   k, maxk, classify-on-add, classify-on-drag, grid, x/y bounds
    //    visibility  controls (all), k-slider, classify-button, toggles
    //  so a figure can pin a behaviour while hiding its control.
    // -----------------------------------------------------------------------

    // k must be odd: with two classes an even k can split the vote, and the
    // tie-break would then decide the label instead of the neighbours.
    const toOdd = (v) => 2 * Math.round((v - 1) / 2) + 1;

    class KnnIllustrator extends P.Widget {
        setup() {
            const d = this.domain({ x: [-1, 1], y: [-0.3, 1.2] });
            this.classifyOnAdd = this.flag("classify-on-add", true);
            this.classifyOnDrag = this.flag("classify-on-drag", true);

            const panel = this.panel = new P.Panel({
                width: 640, height: 400, margin: P.MARGIN,
                x: { domain: d.x }, y: { domain: d.y },
                class: "dp-svg dp-interactive",
                ariaLabel: P.t("knn.aria", "Ilustração interativa de classificação por k vizinhos mais próximos"),
            });

            // Distance is measured in SCREEN space, not data space. The two are
            // not proportional here (the axes have different px-per-unit), so
            // this is a real choice: it guarantees the highlighted neighbours are
            // the ones that *look* nearest, which is what the picture is for.
            // Swap in P.metric.data for true feature-space k-NN.
            this.metric = P.metric.screen(panel);

            this.train = this._seed(50);
            this.queries = [];

            // Clamp k to the odd values the slider can reach, and never past the
            // number of neighbours that actually exist — note toOdd rounds up on
            // a tie, so the ceiling has to round DOWN to stay inside the set.
            const oddFloor = this.train.length % 2 ? this.train.length : this.train.length - 1;
            this.maxk = Math.min(Math.max(1, toOdd(this.int("maxk", 15))), oddFloor);
            this.k = Math.min(Math.max(1, toOdd(this.int("k", 3))), this.maxk);

            panel.drawGrid({ visible: this.flag("grid", true) });
            this.links = panel.segments({ layer: "links", class: "dp-link" });
            this.trainPts = panel.points({
                layer: "train", data: this.train, r: 6,
                attrs: (p) => ({ "data-accent": p.label }),
            });
            this.queryPts = panel.points({
                layer: "queries", data: this.queries, r: 7,
                class: (p) => "dp-point dp-point-query" + (p.label ? "" : " dp-pending"),
                attrs: (p) => ({ "data-accent": p.label }),
            });
            this.queryPts.draggable({
                onMove: (i) => {
                    // A verdict reached at the old position is not a verdict here,
                    // so with live classification off the point goes back to pending
                    // rather than carrying a stale label around the plot.
                    if (this.classifyOnDrag) this._classify(i);
                    else this._unclassify(i);
                    // refresh, not draw: draw would destroy the node holding
                    // pointer capture and the drag would die mid-gesture.
                    this.queryPts.refresh(i);
                    this._drawLinks();
                },
                onRemove: (i) => {
                    this.queries.splice(i, 1);
                    this.queryPts.draw();
                    this._drawLinks();
                },
            });
            panel.background({
                onAdd: (p) => {
                    this.queries.push({ x: p.x, y: p.y, label: null });
                    if (this.classifyOnAdd) this._classify(this.queries.length - 1);
                    this.queryPts.draw();
                    this._drawLinks();
                },
            });

            this.mount(...[panel.svg, this._controls()].filter(Boolean));
            this.trainPts.draw();
        }

        _controls() {
            if (!this.flag("controls", true)) return null;
            const ui = P.controls();

            if (this.flag("k-slider", true)) {
                this.kSlider = ui.slider({
                    min: 1, max: this.maxk, step: 2, value: this.k,
                    label: "k",
                    ariaLabel: P.t("knn.k_aria", "Número de vizinhos"),
                    onInput: (v) => this._setK(v),
                });
            }
            if (this.flag("classify-button", true)) {
                ui.button(P.t("knn.classify", "Classificar pontos"), () => this._classifyAll());
            }
            if (this.flag("toggles", true)) {
                ui.checkbox(P.t("knn.on_add", "classificar ao adicionar"), this.classifyOnAdd,
                    (on) => { this.classifyOnAdd = on; });
                ui.checkbox(P.t("knn.on_drag", "classificar ao arrastar"), this.classifyOnDrag,
                    (on) => { this.classifyOnDrag = on; });
            }
            ui.hint(P.t("knn.hint",
                "Clique no plano para adicionar um ponto · arraste para mover · duplo-clique remove."));
            // Every control individually hidden: drop the bar rather than leave an
            // empty flex row contributing its top margin.
            if (!ui.bar.childNodes.length) ui.bar.remove();
            return ui.el;
        }

        // Two overlapping clouds: class A lower-left, class B upper-right. Both
        // are generated inside the plotted window — a point outside it is one the
        // reader cannot see but the classifier still counts.
        _seed(perClass) {
            const [xmin, xmax] = this.panel.x.domain;
            const [ymin, ymax] = this.panel.y.domain;
            const w = xmax - xmin, h = ymax - ymin;
            const pts = [];
            for (let i = 0; i < perClass; i++) {
                pts.push({
                    x: xmin + Math.random() * w * 0.5,
                    y: ymin + Math.random() * h * 0.4,
                    label: "orange",
                });
                pts.push({
                    x: xmin + w * (0.35 + 0.65 * Math.random()),
                    y: ymin + Math.random() * h * 0.75,
                    label: "purple",
                });
            }
            return pts;
        }

        _classify(i) {
            const q = this.queries[i];
            const ranked = this.train
                .map((p, j) => ({ j, d: this.metric(q, p) }))
                .sort((a, b) => a.d - b.d);

            const k = Math.min(this.k, ranked.length);
            const counts = new Map();
            q.neighbors = [];
            for (let n = 0; n < k; n++) {
                const p = this.train[ranked[n].j];
                q.neighbors.push(p);
                counts.set(p.label, (counts.get(p.label) || 0) + 1);
            }

            // Plain argmax. A tie breaks toward the label seen first, i.e. the
            // one belonging to the nearer neighbour — which is exactly why an
            // odd k is the usual advice for two classes.
            let best = null, bestCount = -1;
            for (const [label, n] of counts) if (n > bestCount) { best = label; bestCount = n; }
            q.label = best;
        }

        // Back to pending: no label, no neighbours, so nothing on screen claims a
        // verdict this point no longer has.
        _unclassify(i) {
            const q = this.queries[i];
            q.label = null;
            q.neighbors = [];
        }

        _classifyAll() {
            this.queries.forEach((_, i) => this._classify(i));
            this.queryPts.draw();
            this._drawLinks();
        }

        // Changing k re-votes every point that already has a verdict, so the
        // boundary visibly shifts as the reader slides. Pending points stay
        // pending — the reader has not asked for those yet.
        _setK(v) {
            this.k = Math.min(Math.max(1, toOdd(v)), this.maxk);
            if (this.kSlider) this.kSlider.value.textContent = this.k;
            this.queries.forEach((q, i) => { if (q.label != null) this._classify(i); });
            this.queryPts.draw();
            this._drawLinks();
        }

        // Rebuild every fan from state. Appending per query point is what used
        // to leak: the lines were detached from the DOM but never from the list.
        _drawLinks() {
            this.links.set(this.queries.flatMap((q) =>
                (q.neighbors || []).map((p) => ({ x1: q.x, y1: q.y, x2: p.x, y2: p.y })),
            ));
        }
    }

    P.define("delta-knn-example", KnnIllustrator);

    // Published so a future package can build on the same toolkit: declare
    // `"needs": ["regression"]` in its manifest and read window.DeltaPlot.
    window.DeltaPlot = window.DeltaPlot || P;
})();
